corehq/ex-submodules/pillowtop/es_utils.py

```diff
@@ -29,6 +29,86 @@
 }
 
 
+def _get_analysis(*names):
+    return {
+        "analyzer": {name: ANALYZERS[name] for name in names}
+    }
+
+
+ANALYZERS = {
+    "default": {
+        "type": "custom",
+        "tokenizer": "whitespace",
+        "filter": ["lowercase"]
+    },
+    "comma": {
+        "type": "pattern",
+        "pattern": r"\s*,\s*"
+    },
+    "sortable_exact": {
+        "type": "custom",
+        "tokenizer": "keyword",
+        "filter": ["lowercase"]
+    },
+}
+
+REMOVE_SETTING = None
+
+ES_ENV_SETTINGS = {
```

Comment:
The actual current configuration for production and icds (looking at ES head) are actually more like this:

```
production:
  default:
    number_of_shards: 5
    number_of_replicas: 1
  hqapps:
    number_of_replicas: 0
  hqusers:
    number_of_shards: 2
icds:
  default:
    number_of_shards: 5
    number_of_replicas: 1
  report_xforms:
    number_of_replicas: 0
  report_cases:
    number_of_replicas: 0
  hqusers:
    number_of_replicas: 2  # stored in triplicate
    number_of_shards: 2
  hqapps:
    number_of_replicas: 0
```

(There are also some unaliased indexes that I've excluded that we can probably delete.)

I'm not sure why hqusers has 2 shards instead of 5, or where this value of 5 shards is getting set on the other indexes, since it doesn't appear here anywhere. The default of 1 replica I'm guessing was manually done using `./scripts/elasticsearch-administer.py create_replica_shards`.

I'm also not sure how to automatically construct a diff between config values and live value, or the exact conditions under which these config values get propagated to the live service, but it seems that this at least happens when a new index is created (including via preindex after "changing" an index).

Reply:
> (There are also some unaliased indexes that I've excluded that we can probably delete.)

I think `./manage.py prune_elastic_indices` should confirm that for you

Reply:
5 is the default in ES. I think 2 came from some amount of optimization and trying to be better about ES setup (I think it was either myself or ben rudolph a few years ago). having more shards is theoretically higher overhead (since a query would have to go 5 machines instead of 2) and users is really small. 

not sure on how necessary that is to keep.

Reply:
Ah okay, if 5 is the es default, that fleshes out the story in my head quite a bit; the simplified version of the block above would then be

```
production:
  default:
    number_of_replicas: 1
  hqapps:
    number_of_replicas: 0
icds:
  default:
    number_of_replicas: 1
  report_xforms:
    number_of_replicas: 0
  report_cases:
    number_of_replicas: 0
  hqusers:
    number_of_replicas: 2  # stored in triplicate
  hqapps:
    number_of_replicas: 0
```

If we further assume as I suggested below that `number_of_replicas: 1` is the default if you have more than one es machine, then we'd use in commcare-cloud to get exactly this is

production:
```
hqapps:
  number_of_replicas: 0
```
icds:
```
  report_xforms:
    number_of_replicas: 0
  report_cases:
    number_of_replicas: 0
  hqusers:
    number_of_replicas: 2  # stored in triplicate
  hqapps:
    number_of_replicas: 0
```

and I would suggest that the hqapps thing is just a mistake, and accident of history, and we'd do well to just exclude the production overrides entirely and use

icds:
```
  report_xforms:
    number_of_replicas: 0
  report_cases:
    number_of_replicas: 0
  hqusers:
    number_of_replicas: 2  # stored in triplicate
```

as the only intentional deviation from the default.